examples/examples_test.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+set +e
+
+clean () {
+    jobs -p | xargs pkill -P 
+    wait
+}
+
+fail () {
+    echo "$(tput setaf 1) $1 $(tput sgr 0)"
+    clean
+    exit 1
+}
+
+pass () {
+    echo "$(tput setaf 2) $1 $(tput sgr 0)"
+}
+
+EXAMPLES=(
+    "helloworld"
+    "route_guide"
+    "features/authentication"
+    "features/compression"
+    "features/deadline"
+    "features/encryption/TLS"
+    )
+
+declare -A EXPECTED_SERVER_OUTPUT=( 
+    ["helloworld"]="Received: world"
+    ["route_guide"]=""
+    ["features/authentication"]="server starting on port 50051..."
+    ["features/compression"]="UnaryEcho called with message \"compress\""
+    ["features/deadline"]=""
+    ["features/encryption/TLS"]=""
+    )
+
+declare -A EXPECTED_CLIENT_OUTPUT=(
+    ["helloworld"]="Greeting: Hello world"
+    ["route_guide"]="location:<latitude:416851321 longitude:-742674555 >"
+    ["features/authentication"]="UnaryEcho:  hello world"
+    ["features/compression"]="UnaryEcho call returned \"compress\", <nil>" 
+    ["features/deadline"]="wanted = DeadlineExceeded, got = DeadlineExceeded" 
+    ["features/encryption/TLS"]="UnaryEcho:  hello world"
+    )
+
+for example in ${EXAMPLES[@]}; do
+    echo "$(tput setaf 4) testing: ${example} $(tput sgr 0)"
+
+    # Build server
+    if ! go build -o /dev/null ./examples/${example}/*server/*.go; then
+        fail "failed to build server"
+    else
+        pass "successfully built server"
+    fi
+
+    # Build client
+    if ! go build -o /dev/null ./examples/${example}/*client/*.go; then
```

Comment:
I think this check makes some sense, especially for the server (since we don't check the result of running the server binary), because it distinguishes the failure modes.  It's not important (the test would fail either way), but it could be helpful.

Reply:
`go build` is used to test that the binary can be built. I outputted the binary into /dev/null so it doesn't have to be cleaned.

`go run` is used to test if the behavior of the example is expected.

We could run the binary instead of using go run also, but i'm not sure if that makes a big difference.